homeassistant/components/climate/__init__.py

```diff
@@ -288,6 +300,102 @@ class ClimateEntity(Entity, cached_properties=CACHED_PROPERTIES_WITH_ATTR_):
+            """Log warning not implemented turn on/off feature."""
+            report_issue = self._suggest_report_issue()
+            if feature.startswith("TURN"):
+                message = (
+                    "Entity %s (%s) does not set ClimateEntityFeature.%s"
+                    " but implements the %s method. Please %s"
+                )
+            else:
+                message = (
+                    "Entity %s (%s) implements HVACMode.%s and therefore implicitly"
+                    " supports the %s service without setting the proper"
+                    " ClimateEntityFeature. Please %s"
+                )
+            _LOGGER.warning(
+                message,
+                self.entity_id,
+                type(self),
+                feature,
+                feature.lower(),
+                report_issue,
+            )
+
+        # Adds ClimateEntityFeature.TURN_OFF/TURN_ON depending on service calls implemented
+        # This should be removed in 2025.1.
+        if not self.supported_features & ClimateEntityFeature.TURN_OFF:
+            if (
+                type(self).async_turn_off is not ClimateEntity.async_turn_off
+                or type(self).turn_off is not ClimateEntity.turn_off
+            ):
+                # turn_off implicitly supported by implementing turn_off method
+                _report_turn_on_off("TURN_OFF", "turn_off")
+                self.__mod_supported_features |= (  # pylint: disable=unused-private-member
+                    ClimateEntityFeature.TURN_OFF
+                )
+            elif HVACMode.OFF in self.hvac_modes:
+                # turn_off implicitly supported by including HVACMode.OFF
+                _report_turn_on_off("off", "turn_off")
+                self.__mod_supported_features |= (  # pylint: disable=unused-private-member
+                    ClimateEntityFeature.TURN_OFF
+                )
+
+        if not self.supported_features & ClimateEntityFeature.TURN_ON:
+            if (
+                type(self).async_turn_on is not ClimateEntity.async_turn_on
+                or type(self).turn_on is not ClimateEntity.turn_on
+            ):
+                # turn_on implicitly supported by implementing turn_on method
+                _report_turn_on_off("TURN_ON", "turn_on")
+                self.__mod_supported_features |= (  # pylint: disable=unused-private-member
+                    ClimateEntityFeature.TURN_ON
+                )
+            else:
+                for _mode in self.hvac_modes:
+                    if _mode != HVACMode.OFF:
+                        # turn_on implicitly supported by including any other HVACMode than HVACMode.OFF
+                        _report_turn_on_off(_mode, "turn_on")
+                        self.__mod_supported_features |= (  # pylint: disable=unused-private-member
+                            ClimateEntityFeature.TURN_ON
+                        )
+                        break
```

Comment:
I think this would be more pythonic, but it's a matter of taste I guess:
```suggestion
                if any _mode != HVACMode.OFF  for _mode in self.hvac_modes:
                    # turn_on implicitly supported by including any other HVACMode than HVACMode.OFF
                    _report_turn_on_off(_mode, "turn_on")
                    self.__mod_supported_features |= (  # pylint: disable=unused-private-member
                        ClimateEntityFeature.TURN_ON
                    )
                    break
```